integrated_channels/moodle/client.py

```diff
@@ -0,0 +1,233 @@
+            response = method(self, *args, **kwargs)
+        elif error_code:
+            raise ClientError(
+                'Moodle API Client Task "{method}" failed with error code '
+                '"{code}" and message: "{msg}" '.format(
+                    method=method.__name__, code=error_code, msg=body.get('message')
+                )
+            )
+        elif warnings:
+            # More Moodle nonsense!
+            errors = []
+            for warning in warnings:
+                if warning.get('message'):
+                    errors.append(warning.get('message'))
+            raise ClientError(
+                'Moodle API Client Task "{method}" failed with the following error codes: '
+                '"{code}"'.format(method=method.__name__, code=errors)
+            )
+        return response
+    return inner
+
+
+class MoodleAPIClient(IntegratedChannelApiClient):
+    """
+    Client for connecting to Moodle.
+    Transmits learner and course metadata.
+
+    Required configuration to access Moodle:
+    - wsusername and wspassword:
+        - Web service user and password created in Moodle. Used to generate api tokens.
+    - Moodle base url.
+        - Customer's Moodle instance url.
+        - For local development just `http://localhost` (unless you needed a different port)
+    - Moodle service short name.
+        - Customer's Moodle service short name
+    """
+
+    MOODLE_API_PATH = '/webservice/rest/server.php'
+
+    def __init__(self, enterprise_configuration):
+        """
+        Instantiate a new client.
+
+        Args:
+            enterprise_configuration (MoodleEnterpriseCustomerConfiguration): An enterprise customers's
+            configuration model for connecting with Moodle
+        """
+        super(MoodleAPIClient, self).__init__(enterprise_configuration)
+        self.config = apps.get_app_config('moodle')
+        self.token = enterprise_configuration.token or self._get_access_token()
+
+    @moodle_request_wrapper
+    def create_course_completion(self, user_id, payload):  # pylint: disable=unused-argument
+        pass
+
+    @moodle_request_wrapper
+    def delete_course_completion(self, user_id, payload):  # pylint: disable=unused-argument
+        pass
+
+    @moodle_request_wrapper
+    def _get_courses(self, key):
+        """
+        Gets courses from Moodle by key (because we cannot update/delete without it).
+        """
+        params = {
+            'wstoken': self.token,
+            'wsfunction': 'core_course_get_courses_by_field',
+            'field': 'shortname',
+            'value': key,
+            'moodlewsrestformat': 'json'
+        }
+        response = requests.get(
+            urljoin(
+                self.enterprise_configuration.moodle_base_url,
+                self.MOODLE_API_PATH,
+            ),
+            params=params
+        )
+        return response
+
+    def get_course_id(self, key):
```

Comment:
small nitpick here- but for consistency, do we want this to mark this as an internal method? Similar to `_get_courses`

Reply:
I specifically made it not an internal method just so I don't have to throw pylint exception comments in a bunch of spots (mostly tests though) since I do test the raw method call.
I'm not really sure there's much value in making any of the classes private to be honest since Python doesn't actually really enforce privacy. I think for get/update/post it makes sense also because we're overriding methods already named get/update/post.